### DISCUSSION TITLE
What's the recommended way to have a proper typed `Props` properties

### DISCUSSION THREAD

**Discussion by 0xbuidler** [lit/lit#4592] [General]
I want my `LitElement` to use my `Props` type as properties. How can I do that ? I cannot see anything in the documentation regarding that.

**Comment by augustjk** [comment]
You can do
```ts
class MyElement extends LitElement implements Props {

}
```

e.g. [playground example](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQge2h0bWwsIGNzcywgTGl0RWxlbWVudH0gZnJvbSAnbGl0JztcbmltcG9ydCB7Y3VzdG9tRWxlbWVudCwgcHJvcGVydHl9IGZyb20gJ2xpdC9kZWNvcmF0b3JzLmpzJztcblxudHlwZSBQcm9wcyA9IHtcbiAgbmFtZTogc3RyaW5nO1xufVxuXG5AY3VzdG9tRWxlbWVudCgnc2ltcGxlLWdyZWV0aW5nJylcbmV4cG9ydCBjbGFzcyBTaW1wbGVHcmVldGluZyBleHRlbmRzIExpdEVsZW1lbnQgaW1wbGVtZW50cyBQcm9wcyB7XG4gIHN0YXRpYyBzdHlsZXMgPSBjc3NgcCB7IGNvbG9yOiBibHVlIH1gO1xuXG4gIC8vIG5vdCBoYXZpbmcgYG5hbWVgIGhlcmUgd291bGQgdHlwZSBlcnJvcnNcbiAgQHByb3BlcnR5KClcbiAgbmFtZSA9ICdTb21lYm9keSc7XG5cbiAgcmVuZGVyKCkge1xuICAgIHJldHVybiBodG1sYDxwPkhlbGxvLCAke3RoaXMubmFtZX0hPC9wPmA7XG4gIH1cbn1cbiJ9LHsibmFtZSI6ImluZGV4Lmh0bWwiLCJjb250ZW50IjoiPCFET0NUWVBFIGh0bWw-XG48aGVhZD5cbiAgPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi9zaW1wbGUtZ3JlZXRpbmcuanNcIj48L3NjcmlwdD5cbjwvaGVhZD5cbjxib2R5PlxuICA8c2ltcGxlLWdyZWV0aW5nIG5hbWU9XCJXb3JsZFwiPjwvc2ltcGxlLWdyZWV0aW5nPlxuPC9ib2R5PlxuIn0seyJuYW1lIjoicGFja2FnZS5qc29uIiwiY29udGVudCI6IntcbiAgXCJkZXBlbmRlbmNpZXNcIjoge1xuICAgIFwibGl0XCI6IFwiXjMuMC4wXCIsXG4gICAgXCJAbGl0L3JlYWN0aXZlLWVsZW1lbnRcIjogXCJeMi4wLjBcIixcbiAgICBcImxpdC1lbGVtZW50XCI6IFwiXjQuMC4wXCIsXG4gICAgXCJsaXQtaHRtbFwiOiBcIl4zLjAuMFwiXG4gIH1cbn0iLCJoaWRkZW4iOnRydWV9XQ)

**Reply by justinfagnani** [reply]
Just be aware that you can't always implement any standalone interface on top of a custom elements because there could be conflicts. You should design your element API to take into account the fact that you extend HTMLElement (and LitElement).

**Comment by justinfagnani** [comment]
The reason why there's no "props" concept in Lit as in React is that custom elements are objects with numerous properties that can be set, so at the limit the "props" type would be the class interface itself. Filtering the interface by `@property() decorated fields wouldn't be correct because there's fundamentally nothing special about them. Users could set non-reactive fields, fields from the DOM like `id` or `title`, or hand-written accessors.
